### PR TITLE
replace externalProvider principalIDs if they exist

### PR DESCRIPTION
**Pull request by kinarashah**
https://github.com/rancher/rancher/issues/12362

### CODE REVIEW

File: pkg/auth/providers/common/usermanager.go [hunk 1]
```diff
@@ -332,3 +342,12 @@ func groupPrincipalPRTB(obj interface{}) ([]string, error) {
 	}
 	return gp, nil
 }
+
+func providerExists(principalIDs []string, provider string) bool {
+	for _, id := range principalIDs {
+		if strings.Contains(id, provider) {
```
Comment (alena1108) [inline pkg/auth/providers/common/usermanager.go]: is it safe to use `contains` here? What are the rules/examples of principalIds vs provider?

File: pkg/auth/providers/common/usermanager.go [hunk 2]
```diff
@@ -84,6 +84,16 @@ func (m *userManager) SetPrincipalOnCurrentUser(apiContext *types.APIContext, pr
 		return nil, err
 	}
 
+	if providerExists(user.PrincipalIDs, principal.Provider) {
+		var principalIDs []string
+		for _, id := range user.PrincipalIDs {
+			if !strings.Contains(id, principal.Provider) {
```
Comment (cjellick) [inline pkg/auth/providers/common/usermanager.go]: FYI, we have a helper: `slice.ContainsString` that does this

File: pkg/auth/providers/common/usermanager.go (hunk 1 above)
Comment (cjellick) [reply pkg/auth/providers/common/usermanager.go]: it would be safer to split on `://` (or just the first `:`) and then do contains

**Comment by cjellick** [comment]
The change fundamentally LGTM. So, whatever you two decide regarding the looping, you have my :+1:

File: pkg/auth/providers/common/usermanager.go [hunk 3]
```diff
@@ -84,6 +84,16 @@ func (m *userManager) SetPrincipalOnCurrentUser(apiContext *types.APIContext, pr
 		return nil, err
 	}
 
+	if providerExists(user.PrincipalIDs, principal.Provider) {
+		var principalIDs []string
```
[the remark this answers is not in the document]
Comment (alena1108) [reply pkg/auth/providers/common/usermanager.go]: We can return an array of the principalIds then, of those matching the provider. It just seems like an overhead - finding the first occurrence just to figure out the existence, then iterate all over again to form the final list :)

**Pull request merged**